### ISSUE TITLE
Provide alternative message fields unpacking

### ISSUE THREAD

**Issue by furiousassault** [2gis/loggo#9]
Currently we process fields of original log message by flattening nested fields structures into flat map.
Generally, this leads to conflict between fields that may be "unpacked" to the top-level dictionary: depending on our logic, loggo-generated or docker-top-level fields may override user' ones or vice versa.

On the other side, we already have to preserve backward compatibility for current users.
So it seems it's needed to implement both variants: current flattening scheme and the one, where initial user log structure would be left untouched under some configurable field name, like `log`.

**Comment by nwlunatic** [comment]
Can you, please, provide current and desired log messages processing result? 

If I understood you right, we currently have  3 sources of fields in log messages:
* user log fields
* docker container injected context (like container_name, etc.)
* loggo injected context (?)

the current possible flatmap of result message may be
```
{
  "container_name": "some-container" // this came from docker
  "container_name": "barrel"  // this is a user field that has nothing to do with docker
}
```
which will result in a conflict, that will eventually be resolved (based on a sequence) in one of them be overwritten by the other.

The solution to this as you suggested is prefixing the source with it's name. Although, I would suggest to prefix additional context information instead of user log fields, so it will result in:
```
{
  "container_name": "barrel" // user field left untouched. But may also be prefixed with user.<field> to fully resolve any possible conflict
  "docker.container_name": "some-container" // we prefixed every docker filed with `docker` prefix
  "loggo.containter_name": "don't make any sense" // we prefixed loggo context with loggo filed
}
```

The reason to do so is that from user point of view his logs have a priority over our additional context information by my opinion

**Comment by furiousassault** [comment]
Thank you for looking over.
In general you've understood the issue right.

In case of docker, we have some log like this: {"log": "user message with nested fields here", "docker field 0": "value" ... }
We do base JSON unmarshal into map called outer map.

Then, we try to unmarshal log field as json, to map[string]interface{}, and recursively flatten it into outer map. If field is nested, its key is generated with dot notation. 
Then we set our loggo fields to the same outer map. Some of user and loggo fields are used in Loggo logic after parsing (SLI gathering) before marshalling map again and sending to a transport, so this logic might be changed as well.

I see two possible solutions:
1. Yours, where we put user fields on the top level (probably with prefix), still flattening them (or not? if I got you right), but pull docker and loggo fields to nested maps, or add prefixes to their names (keys). Seems not bad, because it would not affect current user log mappings.
2. We flatten a copy of user log only for our own purposes (parsing predefined fields, SLI logic, etc), but leave log field of original message untouched, whether is it a map or string. Flattened fields are not put in the result map.

Additional loggo fields are set with prefixes or in map field named "loggo".
So the one of the main questions for me is to flatten something or not, to have flat resulting structure (map[string]interface{} but without nested maps or lists) or arbitrary nested, regarding both the user log and our utility field groups (docker, loggo).

**Comment by furiousassault** [comment]
Current: https://github.com/2gis/loggo#log-fields-transformations (all in one heap). There's no nested fields in user log, but they would appear plainified with prefixes as described).
Desired one is not determined for me yet. As a minimum, it must not break loggo additional logic, and eliminate conflicts.

**Comment by furiousassault** [comment]
Closed by #10 .
